dali/core/mm/pinned_pool_test.cc

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) 2021, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cuda_runtime.h>
+#include <gtest/gtest.h>
+#include <random>
+#include <vector>
+#include "dali/core/mm/async_pool.h"
+#include "dali/core/dev_buffer.h"
+#include "dali/core/mm/mm_test_utils.h"
+#include "dali/core/cuda_stream.h"
+#include "rmm/mr/host/pinned_memory_resource.hpp"
+#include "dali/test/tensor_test_utils.h"
+
+namespace dali {
+namespace mm {
+namespace test {
+
+TEST(MMPinnedAlloc, StageCopy) {
+  test_pinned_resource upstream;
+  {
+    CUDAStream stream;
+    stream = CUDAStream::Create(true);
+    stream_view sv(stream);
+    async_pool_base<memory_kind::pinned> pool(&upstream);
+    std::mt19937_64 rng;
+    const int N = 1<<20;
+    vector<uint8_t> pattern(N), copy_back(N);
+    DeviceBuffer<uint8_t> dev_buf;
+    dev_buf.resize(N);
+    UniformRandomFill(pattern, rng, 0, 255);
+    void *mem1 = pool.allocate(N);
+    memcpy(mem1, pattern.data(), N);
+    CUDA_CALL(cudaMemcpyAsync(dev_buf, mem1, N, cudaMemcpyHostToDevice, stream));
+    pool.deallocate_async(mem1, N, sv);
+    void *mem2 = pool.allocate_async(N, sv);
+    EXPECT_EQ(mem1, mem2);
+    CUDA_CALL(cudaMemcpyAsync(copy_back.data(), dev_buf, N, cudaMemcpyDeviceToHost, stream));
+    CUDA_CALL(cudaStreamSynchronize(stream));
+    pool.deallocate_async(mem1, N, sv);
```

Comment:
Would it make any sense to swap deallocate_async and cudaStreamSynchronize?

Reply:
It makes no difference, since the pool will never truly deallocate the memory, so it's going to remain available anyway.